File: eom/tokens.py
"""Pinned tokeniser for EOM compactness checks.

The harness fixes cl100k_base so that B_A / B_AB budgets are
reproducible across implementations, regardless of the LLM that
ultimately consumes the context pack.
"""

from functools import lru_cache

import tiktoken

ENCODING_NAME = "cl100k_base"


@lru_cache(maxsize=1)
def _encoding() -> tiktoken.Encoding:
    return tiktoken.get_encoding(ENCODING_NAME)


def count_tokens(text: str) -> int:
    """Return the number of tokens in `text` under the pinned encoding.

    Special-token strings such as ``<|endoftext|>`` are tokenised as
    ordinary text rather than raising, because EOM source text may
    legitimately quote LLM control tokens (e.g. articles about LLMs).
    """
    if not text:
        return 0
    return len(_encoding().encode(text, disallowed_special=()))

File: web/src/eom/tokens.ts
/**
 * Pinned tokeniser for EOM compactness checks.
 *
 * Mirrors eom/tokens.py — fixes cl100k_base so B_A / B_AB budgets are
 * reproducible across language implementations, regardless of the LLM that
 * ultimately consumes the context pack. The Python side uses tiktoken;
 * here we use gpt-tokenizer's cl100k bundle, which is byte-exact.
 *
 * Special-token strings such as `<|endoftext|>` are tokenised as ordinary
 * text rather than raising, because source text may legitimately quote
 * LLM control tokens.
 */

import { encode } from "gpt-tokenizer/encoding/cl100k_base";

export const ENCODING_NAME = "cl100k_base";

export function countTokens(text: string): number {
  if (!text) return 0;
  return encode(text).length;
}
